src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,178 @@
+        start.setType(XData::Field::Field_TextSingle);
+        type.setVar("start");
+        d->from.setTimeSpec(Qt::UTC);
+        type.setValue(QStringList(d->from.toString()));
+        fl.append(start);
+    }
+
+    if (d->to.isValid()) {
+        XData::Field end;
+        end.setType(XData::Field::Field_TextSingle);
+        type.setVar("end");
+        d->to.setTimeSpec(Qt::UTC);
+        type.setValue(QStringList(d->to.toString()));
+        fl.append(end);
+    }
+
+    // TODO: ID filters?
+
+    XData x;
+    x.setType(XData::Data_Submit);
+    x.setFields(fl);
+
+    return x;
+}
+
+// TODO: use iris rsm implementation
+void MAMTask::getPage()
+{
+    QDomElement iq = createIQ(doc(), "set", QString(), id());
+
+    QDomElement query = doc()->createElementNS("urn:xmpp:mam:2", "query");
+
+    XData            x = makeMAMFilter();
+
+    QDomElement rsmSet = doc()->createElementNS("http://jabber.org/protocol/rsm", "set");
+    rsmSet.appendChild(textTag(doc(), "max", QString::number(d->mamMaxMessages)));
+
+    if(d->flipPages) rsmSet.appendChild(emptyTag(doc(), "flip-page"));
+    if(d->backwards) {
+        if(d->lastArchiveID.isNull()) {
+            rsmSet.appendChild(emptyTag(doc(), "before"));
+        } else {
+            rsmSet.appendChild(textTag(doc(), "before", d->lastArchiveID));
+        }
+    } else {
+        if(!d->lastArchiveID.isNull()) {
+            rsmSet.appendChild(textTag(doc(), "after", d->lastArchiveID));
+        }
+    }
+
+    if(!(d->lastArchiveID.isNull() || d->lastArchiveID.isEmpty())) {
+        rsmSet.appendChild(textTag(doc(), "after", d->lastArchiveID));
+    }
+
+    query.appendChild(x.toXml(doc()));
+    query.appendChild(rsmSet);
+    iq.appendChild(query);
+    send(iq);
+}
+
+void MAMTask::getArchiveMetadata() {
+    // Craft a query to get the first and last messages in an archive
+    QDomElement iq = createIQ(doc(), "get", QString(), id());
+    QDomElement metadata = emptyTag(doc(), "metadata");
+    metadata.setAttribute("xmlns", "urn:xmpp:mam:2");
+    iq.appendChild(metadata);
+    iq.appendChild(makeMAMFilter().toXml(doc()));
+
+    send(iq);
+}
+
+void MAMTask::onGo() {
+    getArchiveMetadata();
+}
+
+bool MAMTask::take(const QDomElement &x)
+{
+    if(d->metadataFetched) {
+        if(iqVerify(x, QString(), id())) return false;
+
+        d->archive.append(x);
+        d->lastArchiveID = x.attribute("id");
+        d->messagesFetched = d->messagesFetched + 1;
+
+        // Check if we are done
+        if (x.attribute("id") == d->lastID || d->messagesFetched >= d->mamMaxMessages) {
+            setSuccess();
+        } else if (d->messagesFetched % d->mamPageSize == 0) { 
+            getPage();
+        }
+        // TODO: handle server not sending all MAM stuff gracefully
+    } else {
+        if(!iqVerify(x, QString(), id())) return false;
+
+        if(x.elementsByTagName("start").at(0).isNull() || x.elementsByTagName("end").at(0).isNull() ) return false;
```

Comment:
I understand any string literal by default will be interpreted at latin1 in Qt default. but it would be safer to use `QLatin1String` in all such cases. Unless string literal overload exists.

Reply:
When is this necesary? Looking through the iris code, most of the code doesn't use `QLatin1String`

Reply:
most of the code is quite legacy.
The thing is that methods accepting QString will treat string literal as UTF-8.  QString keeps internal data as UTF-16. 
Converting data from UTF-8 to UTF-16 is in general slower than from latin1 when we don't need to check high bits. In case of string comparisons it's even more preferable to use QLatin1String coz there are overloads of comparison function which bypasses string re-encoding completely.
Yes it may be considered negligible in some particular case , but shall we ignore any optimization completely?